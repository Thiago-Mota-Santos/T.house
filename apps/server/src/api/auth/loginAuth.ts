import type { ParameterizedContext } from "koa";
import { UserModel } from "../../modules/user/UserModel";
import { generateToken } from "../../auth/auth";

interface UserBody {
  email: string;
  password: string;
}

export const loginAuth = async (ctx: ParameterizedContext) => {
  const { email, password } = ctx.request.body as UserBody;
  if (!email || !password) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };

    return;
  }

  const user = await UserModel.findOne({
    email,
    removedAt: null,
  });

  if (!user) {
    console.log("has user");
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
  }

  let correctPassword;

  try {
    correctPassword = user?.authenticate(password);
  } catch {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };

    return;
  }

  if (!correctPassword) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };

    return;
  }

  ctx.status = 200;
  ctx.body = {
    message: "User authenticated sucessfully",
    token: generateToken(user),
  };
};
