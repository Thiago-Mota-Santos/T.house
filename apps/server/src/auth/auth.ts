import type { Next, ParameterizedContext } from "koa";
import jwt from "jsonwebtoken";
import type { User } from "../modules/user/UserModel";
import { UserModel } from "../modules/user/UserModel";
import { config } from "../config";

export const auth = async (ctx: ParameterizedContext, next: Next) => {
  const token = ctx.cookies.get("authToken");

  if (!token) {
    ctx.status = 401;
    ctx.body = {
      message: "Unauthorized",
    };
    return;
  }

  try {
    const subToken = token.replace("JWT ", "");
    const decodedToken = jwt.verify(subToken, config.JWT_SECRET);
    const decodedId = decodedToken as unknown as { user: string };

    const user = await UserModel.findOne({ _id: decodedId.user });
    console.log(user);
    if (!user) {
      ctx.status = 401;
      ctx.body = {
        message: "Unauthorized",
      };
      return;
    }

    ctx.user = user;

    await next();
  } catch (e) {
    ctx.status = 500;
    ctx.body = {
      message: "Some error occurred",
    };
  }
};

export const generateToken = (user: User) => {
  return `JWT ${jwt.sign({ user: user._id }, config.JWT_SECRET)}`;
};
