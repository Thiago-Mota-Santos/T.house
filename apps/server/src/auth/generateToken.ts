import jwt from "jsonwebtoken";
import { config } from "../config";
import type { User } from "../modules/user/UserModel";

const JWT_KEY = config.JWT_SECRET;

export const generateToken = (user: User) => {
  // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
  return `JWT ${jwt.sign({ id: user._id }, JWT_KEY)}`;
};
