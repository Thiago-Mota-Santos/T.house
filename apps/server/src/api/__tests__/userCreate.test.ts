/* eslint-disable no-undef */
import { generateToken } from "../../auth/auth";
import { createUser } from "../../modules/fixtures/createUser";
import { apiCallTestCreate } from "../../test/apiCallTest";
import { clearDbAndRestartCounters } from "../../test/clearDatabase";
import { connectMongoose } from "../../test/connectMongoose";
import { disconnectMongoose } from "../../test/disconnectMoongose";
import { sanitizeTestObject } from "../../test/sanitizeTestObject";

beforeAll(connectMongoose);
beforeEach(clearDbAndRestartCounters);
afterAll(disconnectMongoose);

const url = "/api/user";

it("should return error if header is not set", async () => {
  const response = await apiCallTestCreate({
    url,
    payload: {},
    authorization: null,
  });
  expect(response?.status).toBe(500);
  expect(response?.body.message).toBe("Some error occurred");
  expect(response?.body).toMatchSnapshot();
});

it("should return error if user name it was not passed", async () => {
  const user = await createUser({});
  const authorization = generateToken(user);

  const userBody = { email: "test@test.com", password: "123456" };
  const response = await apiCallTestCreate({
    url,
    payload: { userBody },
    authorization,
  });

  expect(response?.status).toBe(400);
  expect(response?.body.message).toBe("User is required");
  expect(response?.body).toMatchSnapshot();
});

it("should return error if user email it was not passed", async () => {
  const user = await createUser({});
  const authorization = generateToken(user);
  const userBoby = { name: "test@test.com", password: "123456" };
  const response = await apiCallTestCreate({
    url,
    payload: { userBoby },
    authorization,
  });

  expect(response?.status).toBe(400);
  expect(response?.body.message).toBe("User is required");
  expect(response?.body).toMatchSnapshot();
});

it("should return error if user password it was not passed", async () => {
  const user = await createUser({});
  const authorization = generateToken(user);

  const userBody = { name: "test", email: "test@test.com" };
  const response = await apiCallTestCreate({
    url,
    payload: { userBody },
    authorization,
  });

  expect(response?.status).toBe(400);
  expect(response?.body.message).toBe("User is required");
  expect(response?.body).toMatchSnapshot();
});

it("should create user with success", async () => {
  const userCreate = await createUser({});
  const authorization = generateToken(userCreate);
  const user = {
    name: "User 21",
    email: "user2@test.com",
    password: "123456",
  };
  const response = await apiCallTestCreate({
    url,
    payload: { user },
    authorization,
  });

  expect(response?.status).toBe(200);
  expect(response?.body.user.name).toBe(user.name);
  expect(response?.body.user.email).toBe(user.email);
  expect(sanitizeTestObject(response?.body)).toMatchSnapshot();
});
