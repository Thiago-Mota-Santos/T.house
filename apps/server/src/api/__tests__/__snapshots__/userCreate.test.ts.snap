// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create user with success 1`] = `
{
  "user": {
    "__v": 0,
    "_id": "ObjectId",
    "createdAt": "FROZEN-CREATEDAT",
    "email": "user2@test.com",
    "name": "User 21",
    "password": "FROZEN-PASSWORD",
    "removedAt": "EMPTY",
    "updatedAt": "FROZEN-UPDATEDAT",
  },
}
`;

exports[`should return error if header is not set 1`] = `
{
  "message": "Some error occurred",
}
`;

exports[`should return error if user email it was not passed 1`] = `
{
  "message": "User is required",
}
`;

exports[`should return error if user name it was not passed 1`] = `
{
  "message": "User is required",
}
`;

exports[`should return error if user password it was not passed 1`] = `
{
  "message": "User is required",
}
`;
